aten/src/ATen/core/Generator.h

```diff
@@ -1,24 +1,95 @@
 #pragma once
 
-#include <ATen/core/ATenGeneral.h>
 #include <stdint.h>
+#include <memory>
+#include <mutex>
+#include <random>
+#include <deque>
+#include <atomic>
+#include <typeinfo>
+#include <utility>
+
+#include "c10/util/Exception.h"
+#include "ATen/core/Backend.h"
+#include "ATen/detail/CUDAHooksInterface.h"
+
+/*
+* Generator note.
+* A Pseudo Random Number Generator (PRNG) is an engine that uses an algorithm to 
+* generate a seemingly random sequence of numbers, that may be later be used in creating 
+* a random distribution. Such an engine almost always maintains a state and requires a
+* seed to start off the creation of random numbers. Often times, users have
+* encountered that it could be beneficial to be able to create, retain, and destroy 
+* generator states and also be able to have control over the seed value.
+*
+* A Generator in ATen gives users the ability to read, write and modify a PRNG engine.
+* For instance, it does so by letting users seed a PRNG engine, get/set the state of the
+* engine, etc.
+*
+* By default, there is one generator state per device, and a device's generator state is 
+* lazily created. A user can use the torch.Generator() api to create their own generator.
+*/
 
 namespace at {
 
+struct Generator;
+
+/*
+* A GeneratorState object contains a generator engine and other state variables.
+* It also has copy and assign constructors so that generator states can be deep copied.
+*/
+struct CAFFE2_API GeneratorState {
+  GeneratorState() = default;
+  ~GeneratorState();
+  GeneratorState(const GeneratorState& other);
+  GeneratorState& operator=(const GeneratorState& other);
```

Comment:
I strongly encourage you to work out how to get the default constructor generated for you automatically. I'm not sure if the fact that you want a copy constructor (I guess, to clone the GeneratorState) means it was wrong to, e.g., make `philox_offset_per_thread` atomic. I'm not sure why it is atomic; it's not explained in the PR description.

Reply:
Fixed by making `philox_offset_per_thread` non atomic. `incrementPhiloxSeed` now returns the current offset value in the std::pair and increments `philox_offset_per_thread` by `+= increment`.